lib/refresh.coffee

```diff
@@ -40,58 +40,105 @@ aliasItem = ($page, $item, oldItem) ->
-          return
-
-  action = if moveWithinPage
-    order = $(this).children().map((_, value) -> $(value).attr('data-id')).get()
-    {type: 'move', order: order}
-  else if moveFromPage
-    console.log 'drag from', $sourcePage.find('h1').text()
-    {type: 'remove'}
-  else if moveToPage
-    $item.data 'pageElement', $thisPage
-    $before = $item.prev('.item')
-    before = getItem($before)
-    item = aliasItem $thisPage, $item, item
-    {type: 'add', item, after: before?.id}
-  action.id = item.id
-  pageHandler.put $thisPage, action
+  destinationIsGhost = $destinationPage.hasClass('ghost')
+
+  moveWithinPage = equals($sourcePage, $destinationPage)
+  moveBetweenDuplicatePages = not moveWithinPage and \
+    $sourcePage.attr('id') == $destinationPage.attr('id')
+
+  if destinationIsGhost or moveBetweenDuplicatePages
+    $(evt.target).sortable('cancel')
+    return
+
+  if moveWithinPage
+    order = getStoryItemOrder($(this))
+    if not _.isEqual(order, $(this).data('order'))
+      pageHandler.put $destinationPage, {id: item.id, type: 'move', order: order}
+    return
+  copying = sourceIsGhost or evt.shiftKey
+  if copying
+    # If making a copy, update the temp clone so it becomes a true copy.
+    $('.shadow-copy').removeClass('shadow-copy')
+      .data($item.data()).attr({'data-id': $item.attr('data-id')})
+  else
+    pageHandler.put $sourcePage, {id: item.id, type: 'remove'}
+  # Either way, record the add to the new page
+  $item.data 'pageElement', $destinationPage
+  $before = $item.prev('.item')
+  before = getItem($before)
+  item = aliasItem $destinationPage, $item, item
+  pageHandler.put $destinationPage,
+                  {id: item.id, type: 'add', item, after: before?.id}
+
+changeMouseCursor = (e, ui) ->
+  $sourcePage = ui.item.data('pageElement')
+  sourceIsGhost = $sourcePage.hasClass('ghost')
+  $destinationPage = ui.placeholder.parents('.page:first')
+  destinationIsGhost = $destinationPage.hasClass('ghost')
+  moveWithinPage = equals($sourcePage, $destinationPage)
+  moveBetweenDuplicatePages = not moveWithinPage and \
+    $sourcePage.attr('id') == $destinationPage.attr('id')
+  copying = sourceIsGhost or (e.shiftKey and not moveWithinPage)
+  if destinationIsGhost or moveBetweenDuplicatePages
+    $('body').css('cursor', 'no-drop')
+    $('.shadow-copy').hide()
+  else if copying
+    $('body').css('cursor', 'copy')
+    $('.shadow-copy').show()
+  else
+    $('body').css('cursor', 'move')
+    $('.shadow-copy').hide()
 
 initDragging = ($page) ->
+  origCursor = $('body').css('cursor')
   options =
     connectWith: '.page .story'
     placeholder: 'item-placeholder'
     forcePlaceholderSize: true
+    delay: 150
   $story = $page.find('.story')
-  $story.sortable(options).on('sortupdate', handleDragging)
+  $story.sortable(options)
+    .on 'sortstart', (e, ui) ->
+      $item = ui.item
+      $story = $item.parents '.story:first'
+      $story.data 'order', getStoryItemOrder($story)
```

Comment:
Wold it work to make this a local variable above line 121 instead of storing this order on the element? It would then be passed into handleDrop (which seems like it would be better to call in the same 'sortstop' call as everything else instead of having two).

Reply:
The second sortstop is needed in order to avoid more complex logic in handleDrop. The shadow-copy needs to be removed in all cases other than a copy. To do that is a single method would require us to add the same line before every return in handleDrop. Having a second sortstop handler is simpler / results in fewer lines of code.

I'll try putting the original story order in a local variable. If that works, we could safely avoid doing anything on sortstop. I put the cleanup here in order to try to avoid the chance of something else referencing it after the drop. Using a closure would avoid that concern.

Reply:
Commit 10048c4 switches to using a closure to pass the original order to handleDrop.

Reply:
Looks good.

I agree you shouldn't try to overload handleDrop like that, I meant that you could move the `handleDrop(e, ui, originalOrder)` call down to the top of the next `sortstop` handler. It wouldn't change how anything works, but it makes it more explicit that the handleDrop call needs to happen just before the shadow is removed, and makes it less likely for someone to mess it up when working on it later than having it in a separate handler.

Reply:
Ah. That makes sense. 8e5e38e